### PR TITLE
do out of source builds in hashed directories

### PR DESCRIPTION
**Pull request by becker33**

### CODE REVIEW

**Comment by adamjstewart** [comment]
What's the motivation for this?

**Review by alalazo — APPROVED** [review]
Is this to permit several build directories from the same source when using `spack dev-build`?

**Comment by rblake-llnl** [comment]
Yes, I think that's the motivation.

File: lib/spack/spack/build_systems/cmake.py [hunk 1]
```diff
@@ -314,7 +314,8 @@ def build_directory(self):
 
         :return: directory where to build the package
         """
-        return os.path.join(self.stage.path, 'spack-build')
+        dirname = 'spack-build=%s' % self.spec.dag_hash(7)
```
Comment (alalazo) [inline lib/spack/spack/build_systems/cmake.py:317]: I didn't notice the `=` sign on the directory name. Can we maybe use an `-` instead:
```suggestion
        dirname = 'spack-build-%s' % self.spec.dag_hash(7)
```
?

**Comment by haampie** [comment]
Hm, this feels a bit backward... it breaks my CI script which uses `spack dev-build ...` and then `cd spack-build`, which no longer exists.

Can't we make the user provide a directory for `spack dev-build` instead? E.g. `spack dev-build -f path/to/build/folder ...`?

Also, the typical use of `cmake` is as follows:

```
git clone sources
cd sources
mkdir build
cd build
cmake ..
```

To me it would feel more natural to be able to do

```
cd build
spack dev-build .. [spec]
```
too.

---

One more comment: I just got back the old behavior of `spack dev-build` by setting `build_directory = 'spack-build'` in my package.py. I noted there's 45 CMakePackage recipes specifying `build_directory` already. If the intention is to allow for multiple build dirs in `spack dev-build`, I guess this won't work for those 45 packages?

**Pull request merged**